### PR TITLE
Workaround API consistency bugs for various resources

### PR DESCRIPTION
**Pull request by manicminer**
Add some helper funcs for post-deletion check and identifying key/password credentials

Increase the default retry limit by one - making it 9 retries after the first attempt for a total of 10 requests with exponential backoff

Consistency fixes for various resources:

- `azuread_application_certificate`: check for consistency on deletion
- `azuread_application_password`: check for consistency on deletion
- `azuread_directory_role_member`: check for consistency on deletion
- `azuread_group`: check for consistency on creation by attempting to patch the new group, perform consistency check on deletion
- `azuread_group_member`: check for consistency on deletion
- `azuread_invitation`: check for consistency on creation by attempting to patch the guest user, use helper func for consistency check on deletion, move flatten/expand funcs to own source file
- `azuread_service_principal`: check for consistency on creation by attempting to patch the new service principal, perform consistency check on deletion
- `azuread_service_principal_certificate`: check for consistency on deletion
- `azuread_service_principal_password`: check for consistency on deletion
- `azuread_user`: check for consistency on deletion

Note that some of these resources already attempted to PATCH themselves after creation with an empty object. It's become clear that the API is filtering these out and returning early, unless there is an actual property in the body to be changed. So now we're choosing a property to update for each resource and ensuring that a valid change is sent in the PATCH request.

Closes: #578 
Closes: #611

### CODE REVIEW

File: internal/services/directoryroles/directory_role_member_resource.go [hunk 1]
```diff
@@ -174,5 +175,21 @@ func directoryRoleMemberResourceDelete(ctx context.Context, d *schema.ResourceDa
 		return tf.ErrorDiagF(err, "Removing member %q from directory role with object ID: %q", id.MemberId, id.DirectoryRoleId)
 	}
 
+	// Wait for membership link to be deleted
+	err = helpers.WaitForDeletion(ctx, func(ctx context.Context) (*bool, error) {
+		client.BaseClient.DisableRetries = true
+		_, status, err := client.GetMember(ctx, id.DirectoryRoleId, id.MemberId)
+		if err != nil {
```
Comment (katbyte) [inline internal/services/directoryroles/directory_role_member_resource.go:182]: can merge these
```suggestion
		if _, status, err := client.GetMember(ctx, id.DirectoryRoleId, id.MemberId); err != nil {
```

File: internal/services/groups/group_member_resource.go [hunk 2]
```diff
@@ -164,5 +165,21 @@ func groupMemberResourceDelete(ctx context.Context, d *schema.ResourceData, meta
 		return tf.ErrorDiagF(err, "Removing member %q from group with object ID: %q", id.MemberId, id.GroupId)
 	}
 
+	// Wait for membership link to be deleted
+	err = helpers.WaitForDeletion(ctx, func(ctx context.Context) (*bool, error) {
+		client.BaseClient.DisableRetries = true
+		_, status, err := client.GetMember(ctx, id.GroupId, id.MemberId)
+		if err != nil {
```
Comment (katbyte) [inline internal/services/groups/group_member_resource.go:172]: and here

File: internal/services/groups/group_resource.go [hunk 3]
```diff
@@ -754,17 +767,34 @@ func groupResourceRead(ctx context.Context, d *schema.ResourceData, meta interfa
 
 func groupResourceDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	client := meta.(*clients.Client).Groups.GroupsClient
+	groupId := d.Id()
 
-	_, status, err := client.Get(ctx, d.Id(), odata.Query{})
+	_, status, err := client.Get(ctx, groupId, odata.Query{})
 	if err != nil {
```
Comment (katbyte) [inline internal/services/groups/group_resource.go:773]: and here

File: internal/services/invitations/invitation_resource.go [hunk 4]
```diff
@@ -206,71 +234,20 @@ func invitationResourceDelete(ctx context.Context, d *schema.ResourceData, meta
...
-			user, status, err := client.Get(ctx, userID, odata.Query{})
-			if err != nil {
-				if status == http.StatusNotFound {
-					return "stub", "Deleted", nil
-				}
-				return nil, "Error", fmt.Errorf("retrieving Invited User with object ID %q: %+v", userID, err)
+	err = helpers.WaitForDeletion(ctx, func(ctx context.Context) (*bool, error) {
+		client.BaseClient.DisableRetries = true
+		_, status, err := client.Get(ctx, userID, odata.Query{})
+		if err != nil {
```
Comment (katbyte) [inline internal/services/invitations/invitation_resource.go:240]: and here

File: internal/services/serviceprincipals/service_principal_certificate_resource.go [hunk 5]
```diff
@@ -301,5 +293,22 @@ func servicePrincipalCertificateResourceDelete(ctx context.Context, d *schema.Re
...
 	}
 
+	// Wait for service principal certificate to be deleted
+	err = helpers.WaitForDeletion(ctx, func(ctx context.Context) (*bool, error) {
+		client.BaseClient.DisableRetries = true
+		servicePrincipal, _, err := client.Get(ctx, id.ObjectId, odata.Query{})
+		if err != nil {
+			return nil, err
+		}
+		credential := helpers.GetKeyCredential(servicePrincipal.KeyCredentials, id.KeyId)
```
Comment (katbyte) [inline internal/services/serviceprincipals/service_principal_certificate_resource.go:303]:
```suggestion
		}

		credential := helpers.GetKeyCredential(servicePrincipal.KeyCredentials, id.KeyId)
```

File: internal/services/serviceprincipals/service_principal_password_resource.go [hunk 6]
```diff
@@ -259,5 +250,22 @@ func servicePrincipalPasswordResourceDelete(ctx context.Context, d *schema.Resou
...
 	}
 
+	// Wait for service principal password to be deleted
+	err = helpers.WaitForDeletion(ctx, func(ctx context.Context) (*bool, error) {
+		client.BaseClient.DisableRetries = true
+		servicePrincipal, _, err := client.Get(ctx, id.ObjectId, odata.Query{})
+		if err != nil {
+			return nil, err
+		}
+		credential := helpers.GetPasswordCredential(servicePrincipal.PasswordCredentials, id.KeyId)
```
Comment (katbyte) [inline internal/services/serviceprincipals/service_principal_password_resource.go:260]:
```suggestion
		}

		credential := helpers.GetPasswordCredential(servicePrincipal.PasswordCredentials, id.KeyId)
```

File: internal/services/serviceprincipals/service_principal_resource.go [hunk 7]
```diff
@@ -642,20 +654,39 @@ func servicePrincipalResourceRead(ctx context.Context, d *schema.ResourceData, m
 
 func servicePrincipalResourceDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	client := meta.(*clients.Client).ServicePrincipals.ServicePrincipalsClient
+	servicePrincipalId := d.Id()
 
-	_, status, err := client.Get(ctx, d.Id(), odata.Query{})
+	_, status, err := client.Get(ctx, servicePrincipalId, odata.Query{})
 	if err != nil {
```
Comment (katbyte) [inline internal/services/serviceprincipals/service_principal_resource.go:660]: and here

File: internal/services/users/user_resource.go [hunk 8]
```diff
@@ -676,19 +677,36 @@ func userResourceRead(ctx context.Context, d *schema.ResourceData, meta interfac
 
 func userResourceDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	client := meta.(*clients.Client).Users.UsersClient
+	userId := d.Id()
 
-	_, status, err := client.Get(ctx, d.Id(), odata.Query{})
+	_, status, err := client.Get(ctx, userId, odata.Query{})
 	if err != nil {
```
Comment (katbyte) [inline internal/services/users/user_resource.go:683]: and here

File: internal/services/users/user_resource.go [hunk 9]
```diff
@@ -676,19 +677,36 @@ func userResourceRead(ctx context.Context, d *schema.ResourceData, meta interfac
...
+			return tf.ErrorDiagPathF(fmt.Errorf("User was not found"), "id", "Retrieving user with object ID %q", userId)
 		}
 
-		return tf.ErrorDiagPathF(err, "id", "Retrieving user with object ID %q", d.Id())
+		return tf.ErrorDiagPathF(err, "id", "Retrieving user with object ID %q", userId)
 	}
 
-	status, err = client.Delete(ctx, d.Id())
+	status, err = client.Delete(ctx, userId)
 	if err != nil {
```
Comment (katbyte) [inline internal/services/users/user_resource.go:692]: and here?

File: internal/services/applications/application_certificate_resource.go [hunk 10]
```diff
@@ -301,5 +293,22 @@ func applicationCertificateResourceDelete(ctx context.Context, d *schema.Resourc
...
 	}
 
+	// Wait for application certificate to be deleted
+	err = helpers.WaitForDeletion(ctx, func(ctx context.Context) (*bool, error) {
+		client.BaseClient.DisableRetries = true
+		app, _, err := client.Get(ctx, id.ObjectId, odata.Query{})
+		if err != nil {
+			return nil, err
+		}
+		credential := helpers.GetKeyCredential(app.KeyCredentials, id.KeyId)
```
Comment (katbyte) [inline internal/services/applications/application_certificate_resource.go:303]:
```suggestion
		}

		credential := helpers.GetKeyCredential(app.KeyCredentials, id.KeyId)
```

File: internal/services/applications/application_password_resource.go [hunk 11]
```diff
@@ -279,5 +270,22 @@ func applicationPasswordResourceDelete(ctx context.Context, d *schema.ResourceDa
...
 	}
 
+	// Wait for application password to be deleted
+	err = helpers.WaitForDeletion(ctx, func(ctx context.Context) (*bool, error) {
+		client.BaseClient.DisableRetries = true
+		app, _, err := client.Get(ctx, id.ObjectId, odata.Query{})
+		if err != nil {
+			return nil, err
+		}
+		credential := helpers.GetPasswordCredential(app.PasswordCredentials, id.KeyId)
```
Comment (katbyte) [inline internal/services/applications/application_password_resource.go:280]:
```suggestion
		}

		credential := helpers.GetPasswordCredential(app.PasswordCredentials, id.KeyId)
```

File: internal/services/conditionalaccess/named_location_resource.go [hunk 12]
```diff
@@ -273,160 +274,51 @@ func namedLocationResourceRead(ctx context.Context, d *schema.ResourceData, meta
 
 func namedLocationResourceDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	client := meta.(*clients.Client).ConditionalAccess.NamedLocationsClient
+	namedLocationId := d.Id()
 
 	if _, ok := d.GetOk("ip"); ok {
-		_, status, err := client.GetIP(ctx, d.Id(), odata.Query{})
+		_, status, err := client.GetIP(ctx, namedLocationId, odata.Query{})
 		if err != nil {
```
Comment (katbyte) [inline internal/services/conditionalaccess/named_location_resource.go:281]: and here

File: internal/services/conditionalaccess/named_location_resource.go [hunk 13]
```diff
@@ -273,160 +274,51 @@ func namedLocationResourceRead(ctx context.Context, d *schema.ResourceData, meta
...
 
-			return tf.ErrorDiagPathF(err, "id", "Retrieving named location with ID %q", d.Id())
+			return tf.ErrorDiagPathF(err, "id", "Retrieving named location with ID %q", namedLocationId)
 		}
 	}
 
 	if _, ok := d.GetOk("country"); ok {
-		_, status, err := client.GetCountry(ctx, d.Id(), odata.Query{})
+		_, status, err := client.GetCountry(ctx, namedLocationId, odata.Query{})
 		if err != nil {
```
Comment (katbyte) [inline internal/services/conditionalaccess/named_location_resource.go:293]: and here

File: internal/services/conditionalaccess/named_location_resource.go [hunk 14]
```diff
@@ -273,160 +274,51 @@ func namedLocationResourceRead(ctx context.Context, d *schema.ResourceData, meta
...
-		Timeout:                   time.Until(timeout),
-		MinTimeout:                5 * time.Second,
-		ContinuousTargetOccurence: 5,
-		Refresh: func() (interface{}, string, error) {
-			client.BaseClient.DisableRetries = true
-			_, status, err := client.Get(ctx, d.Id(), odata.Query{})
+	err = helpers.WaitForDeletion(ctx, func(ctx context.Context) (*bool, error) {
+		client.BaseClient.DisableRetries = true
+		_, status, err := client.Get(ctx, namedLocationId, odata.Query{})
+		if err != nil {
```
Comment (katbyte) [inline internal/services/conditionalaccess/named_location_resource.go:311]: and here

**Review by katbyte — APPROVED** [review]
Thanks @manicminer - LGTM aside from some minor comments i've left inline 🌵

**Comment by manicminer** [comment]
Thanks @katbyte! Have tidied those up 👍

**Comment by manicminer** [comment]
Test results (failures unrelated)

Screenshot 2021-11-12 at 00 33 09

Screenshot 2021-11-12 at 00 34 07

Screenshot 2021-11-12 at 00 33 57

Screenshot 2021-11-12 at 00 33 46

Screenshot 2021-11-12 at 00 33 36

Screenshot 2021-11-12 at 00 33 22

**Pull request merged**